tests/integration/tier0/non-destructive/kernel/test_custom_kernel.py

```diff
@@ -67,37 +57,55 @@ def custom_kernel(shell, hybrid_rocky_image):
     Remove the current installed kernel and install the machine default kernel
     after the test.
     """
+    repo_from_path = _cross_vendor_kernel()
+    custom_kernel_installed = None
     if os.environ["TMT_REBOOT_COUNT"] == "0":
-
-        assert shell("yum install %s -y" % CUSTOM_KERNEL).returncode == 0
-
-        assert shell("grub2-set-default '%s'" % GRUB_SUBSTRING).returncode == 0
-
+        yum_call = f"yum --disablerepo=* --repofrompath=customkernelrepo,{repo_from_path}"
+
+        # Query for all kernels in the repoquery
+        # grep -v exclude any .src packages
+        # rpmdev-sort sort packages by version
+        # tail -1 the last in the list as the latest one
+        #   (we don't really care about the version, install the latest to mitigate any potential dependency issues)
+        available_kernel_version = shell(
+            f"{yum_call} repoquery kernel | grep -v '.src' | rpmdev-sort | tail -1"
+        ).output.strip()
+
+        # Install the kernel from the path provided by the _cross_vendor_kernel
+        # This way we don't rely on any specific version of kernel hardcoded and install what's available in the repository
+        # Disable all other repositories
+        # Call without the gpg check, so we won't need to import the GPG key
+        assert shell(f"{yum_call} install -y --nogpgcheck {available_kernel_version}").returncode == 0
+
+        custom_kernel_installed = re.sub("kernel-(?:\d+:)?", "", available_kernel_version)
+        # Assemble the full title of the custom kernel and set it as default to boot to
+        grub_substring = _get_full_kernel_title(shell, kernel=custom_kernel_installed)
+        assert shell(f"grub2-set-default '{grub_substring}'").returncode == 0
         shell("tmt-reboot -t 600")
 
     yield
 
     if os.environ["TMT_REBOOT_COUNT"] == "1":
-        # Remove the current installed kernel and install the machine default kernel.
-        custom_kernel_release = CUSTOM_KERNEL.rsplit("/", 1)[-1].replace(".rpm", "")
-        assert shell("rpm -e %s" % custom_kernel_release).returncode == 0
-
-        original_kernel = os.popen("rpm -q --last kernel | head -1 | cut -d ' ' -f1").read()
-        original_kernel_release = original_kernel.rsplit("/")[-1].replace(".rpm", "").split("-")[-1]
-
+        # This is kind of naive, but we assume the second latest installed kernel is the original one
+        # We use head to filter the first two lines of the output and tail to filter the bottom line
+        original_kernel = os.popen("rpm -q --last kernel | head -2 | tail -1 | cut -d ' ' -f1").read().strip()
+        original_kernel_title = _get_full_kernel_title(shell, kernel=original_kernel.replace("kernel-", ""))
```

Comment:
Can't we save the original kernel to a variable before the reboot? I mean somewhere before the if statement (if os.environ["TMT_REBOOT_COUNT"] == "0":)? If not, I suggest creating a tmp file on the system that would hold this information

Reply:
That makes sense.